core/server/src/main/java/alluxio/master/AbstractMaster.java

```diff
@@ -240,19 +240,19 @@ protected void flushJournal() {
     }
   }
 
-  protected long appendJournalEntry(JournalEntry entry) {
+  protected void appendJournalEntry(JournalEntry entry, JournalContext journalContext) {
     Preconditions.checkNotNull(mAsyncJournalWriter, PreconditionMessage.ASYNC_JOURNAL_WRITER_NULL);
-    return mAsyncJournalWriter.appendEntry(entry);
+    journalContext.setFlushCounter(mAsyncJournalWriter.appendEntry(entry));
   }
 
   /**
    * Waits for the flush counter to be flushed to the journal. If the counter is
    * {@link AsyncJournalWriter#INVALID_FLUSH_COUNTER}, this is a noop.
    *
-   * @param counter the flush counter
+   * @param journalContext the journal context
    */
-  protected void waitForJournalFlush(long counter) {
-    if (counter == AsyncJournalWriter.INVALID_FLUSH_COUNTER) {
+  private void waitForJournalFlush(JournalContext journalContext) {
+    if (journalContext.getFlushCounter() == AsyncJournalWriter.INVALID_FLUSH_COUNTER) {
```

Comment:
this method is only called from JournalContext.close() ... I don't think it is necessary to try to be defensive in the case someone is using the JournalContext after they closed it, because if that happens there will be other issues (namely, the journaled events will actually not get persisted)